File: user/app1/projects/Project1/Project1Extension/Common/Utility/Logger.swift
// enum LogKind {
//   case log
//   case mark
//   case warn
//   case error
//   case mute
//   case unmute
// }
import Foundation

public struct LogItem {
  let timestamp: Double  //ms from epoch
  let subSystem: String
  let kind: String
  let message: String
}

//00:00:00.000
func formatTimestamp(_ timestamp: Double) -> String {
  let date = Date(timeIntervalSince1970: timestamp / 1000)
  let formatter = DateFormatter()
  formatter.dateFormat = "HH:mm:ss.SSS"
  return formatter.string(from: date)
}

let subSystemIcons: [String: String] = [
  "host": "🟣",
  "ext": "🔸",
  "ui": "🔹",
  "dsp": "🔺",
]

let logKindIcons: [String: String] = [
  "log": "",
  "mark": "🔽",
  "warn": "⚠️",
  "error": "📛",
]

#if DEBUG
  public class LoggerCore {
    let udpLogger = UDPLogger()

    public func pushLogItem(_ item: LogItem) {

      let ts = formatTimestamp(item.timestamp)
      let ssIcon = subSystemIcons[item.subSystem] ?? ""
      let kindIcon = logKindIcons[item.kind] ?? ""

      let logLine = "\(ts) [\(ssIcon)\(item.subSystem)] \(kindIcon) \(item.message)"
      if true {
        print(logLine)
      }
      if true {
        udpLogger.pushLogItem(item)
      }
    }
  }
  public let loggerCore = LoggerCore()

  public class LoggerEntry {

    private let subSystem: String

    public init(subSystem: String) {
      self.subSystem = subSystem
    }

    private func pushLog(_ kind: String, _ message: String) {
      loggerCore.pushLogItem(
        LogItem(
          timestamp: Date().timeIntervalSince1970 * 1000, subSystem: subSystem, kind: kind,
          message: message
        ))
    }

    public func log(_ message: String) {
      pushLog("log", message)
    }

    public func mark(_ message: String) {
      pushLog("mark", message)
    }

    public func warn(_ message: String) {
      pushLog("warn", message)
    }

    public func error(_ message: String) {
      pushLog("error", message)
    }
  }

  public let logger = LoggerEntry(subSystem: "ext")

#else

  public class LoggerCore {
    public init() {}
    public func pushLogItem(_ item: LogItem) {
    }
  }
  public let loggerCore = LoggerCore()

  public class LoggerEntry {
    public init() {}
    public func log(_ message: String) {
    }
    public func mark(_ message: String) {
    }
    public func warn(_ message: String) {
    }
    public func error(_ message: String) {
    }
  }
  public let logger = LoggerEntry()
#endif

File: templates/_common/frontend/src/bridge/logger.ts
const isDebug = location.search.includes("debug=1");
const isDirectLoggingEnabled = location.search.includes("dlog=1");

const loggerOptions = {
  outputToConsole: false,
  sendToWebViewOwner: false,
  sendToLocalHttpLogServer: false,
};
if (isDebug) {
  Object.assign(loggerOptions, {
    outputToConsole: true,
    sendToWebViewOwner: true,
    sendToLocalHttpLogServer: isDirectLoggingEnabled,
  });
}

let postFailed = false;

type LogKind = "log" | "mark" | "warn" | "error" | "mute" | "unmute";

type LogItem = {
  timestamp: number;
  subsystem: string;
  logKind: LogKind;
  message: string;
};

function createLogFormatter() {
  // const subsystemIcons: Record<string, string> = {
  //   host: "🟣",
  //   ext: "🔸",
  //   ui: "🔹",
  //   dsp: "🔺",
  // };

  const logKindIcons: Record<string, string> = {
    log: "",
    mark: "🔽",
    warn: "⚠️",
    error: "📛",
  };

  function formatTimestamp(timestamp: number) {
    const date = new Date(timestamp);
    const hours = String(date.getHours()).padStart(2, "0");
    const minutes = String(date.getMinutes()).padStart(2, "0");
    const seconds = String(date.getSeconds()).padStart(2, "0");
    const milliseconds = String(date.getMilliseconds()).padStart(3, "0");
    //00:00:00.000
    return `${hours}:${minutes}:${seconds}.${milliseconds}`;
  }

  function formatLogLine(logItem: LogItem): string {
    // const ssIcon = subsystemIcons[logItem.subsystem ?? ""] ?? "";
    const kindIcon = logKindIcons[logItem.logKind ?? ""] ?? "";
    const ts = formatTimestamp(logItem.timestamp);
    return `${ts} ${kindIcon} ${logItem.message}`;
  }
  return { formatLogLine };
}
const logFormatter = createLogFormatter();

async function loggingViaLocalHttp(logItem: LogItem) {
  if (!postFailed) {
    try {
      await fetch("http://localhost:9002", {
        method: "POST",
        body: JSON.stringify(logItem),
      });
    } catch {
      //skip sending logs after first failure
      console.log(`failed to post local http log`);
      postFailed = true;
    }
  }
}

function sendLogItemToWebViewOwner(logItem: LogItem) {
  const globalThisTyped = globalThis as unknown as {
    webkit?: {
      messageHandlers: {
        pluginEditor?: {
          postMessage: (msg: {
            type: "log";
            timestamp: number;
            logKind: string;
            message: string;
          }) => void;
        };
      };
    };
  };
  globalThisTyped.webkit?.messageHandlers.pluginEditor?.postMessage({
    type: "log",
    timestamp: logItem.timestamp,
    logKind: logItem.logKind,
    message: logItem.message,
  });
}

function writeLogItemToConsole(logItem: LogItem) {
  const logLine = logFormatter.formatLogLine(logItem);
  console.log(logLine);
}

type LogArguments = (
  | string
  | number
  | boolean
  | object
  | Array<string | number | boolean | object>
  | unknown
)[];

function mapLogArgumentsToString(args: LogArguments) {
  return args
    .map((arg) => {
      if (typeof arg === "object") {
        return JSON.stringify(arg);
      }
      return String(arg);
    })
    .join(" ");
}

function createLoggerEntry() {
  function pushLog(kind: LogKind, args: LogArguments) {
    if (
      !loggerOptions.outputToConsole &&
      !loggerOptions.sendToLocalHttpLogServer &&
      !loggerOptions.sendToWebViewOwner
    ) {
      return;
    }
    const msg = mapLogArgumentsToString(args);

    const logItem: LogItem = {
      timestamp: Date.now(),
      subsystem: "ui",
      logKind: kind,
      message: msg,
    };

    if (loggerOptions.outputToConsole) {
      writeLogItemToConsole(logItem);
    }

    if (loggerOptions.sendToLocalHttpLogServer) {
      void loggingViaLocalHttp(logItem);
    }

    if (loggerOptions.sendToWebViewOwner) {
      sendLogItemToWebViewOwner(logItem);
    }
  }

  return {
    log(...args: LogArguments) {
      pushLog("log", args);
    },
    mark(...args: LogArguments) {
      pushLog("mark", args);
    },
    warn(...args: LogArguments) {
      pushLog("warn", args);
    },
    error(...args: LogArguments) {
      pushLog("error", args);
    },
  };
}

export const logger = createLoggerEntry();
